common/src/main/java/org/conscrypt/Hkdf.java

```diff
@@ -0,0 +1,231 @@
+ * Each invocation of expand or extract uses a new Mac instance and so instances
+ * of Hkdf are thread-safe.
+ */
+public final class Hkdf {
+    // HMAC algorithm to use.
+    private final String hmacName;
+    // Provider to use for getting Mac instances
+    private final Provider provider;
+
+    // Output length for the Mac in use.
+    private final int macLength;
+
+    /**
+     * Creates an Hkdf instance which will use hmacName as the name for the underlying
+     * HMAC algorithm.
+     * <p>
+     * The Hkdf instace will try and obtain Mac instances from an installed Conscrypt Provider with
+     * the same name as this Conscrypt instance would use. Failing that it will look for
+     * different flavous of Conscrypt that have been installed as Providers, and failing that
+     * it will instantiate a private instance of this Conscrypt.
+     *
+     * @param hmacName the name of the HMAC algorithm to use
+     *
+     * @throws NoSuchAlgorithmException if hmacName is not a valid HMAC name
+     */
+    public Hkdf(String hmacName) throws NoSuchProviderException, NoSuchAlgorithmException {
+        this(hmacName, (String) null);
+    }
+
+    /**
+     * Creates an Hkdf instance which will use hmacName as the name for the underlying
+     * HMAC algorithm to obtain from the specified Provider.
+     * <p>
+     * The Hkdf instance will try and obtain Mac instances from an installed Provider with
+     * the name providerName.
+     *
+     * @param hmacName the name of the HMAC algorithm to use
+     * @param providerName the name of the Provider to use
+     * @throws NoSuchProviderException if an invalid Provider name was supplied
+     * @throws NoSuchAlgorithmException if hmacName is not a valid HMAC name
+     */
+    public Hkdf(String hmacName, String providerName)
+        throws NoSuchProviderException, NoSuchAlgorithmException {
+        this(hmacName, findProvider(providerName));
+    }
+
+    /**
+     * Creates an Hkdf instance which will use hmacName as the name for the underlying
+     * HMAC algorithm to obtain from the specified Provider.
+     * <p>
+     * The Hkdf instance will try and obtain Mac instances from the specified Provider which
+     * does not need to be installed as a JVM-wide Provider.
+     *
+     * @param hmacName the name of the HMAC algorithm to use
+     * @param provider the Provider to use
+     * @throws NoSuchAlgorithmException if hmacName is not a valid HMAC name
+     */
+    public Hkdf(String hmacName, Provider provider) throws NoSuchAlgorithmException {
+        Objects.requireNonNull(hmacName);
+        Objects.requireNonNull(provider);
+        this.hmacName = hmacName;
+        this.provider = provider;
+
+        // Stash the MAC length with the bonus that we'll fail fast here if no such algorithm.
+        macLength = Mac.getInstance(hmacName, provider).getMacLength();
+    }
+
+    private static Provider findProvider(String requestedName) throws NoSuchProviderException {
```

Comment:
Oh, is it just to trigger `NoSuchProviderException` earlier? If so, that probably needs a comment at least. (Although, yeesh, is that really what it takes to use JCA?)

It looks like `Mac` has a `clone` method, which would have been another thread-safety option, but Conscrypt doesn't implement it, so we probably can't rely on it. Though, separately, maybe it makes sense to add that to Conscrypt to align with the RI's HMAC.

Reply:
I shall have a think, we can probably do better here.

In an ideal World, the API would be something like Kdf.getInstance("HKDF", <provider>) and the instance would know its own provider and could bind tightly to it whether it's Conscrypt or not.  But that would entail teaching java.security.Provider about a new type of algorithm, which will happen eventually upstream but not for a while.

We could probably align better with the future by:
* Making the constructor non-public
* Having a static getInstance method which searches installed Providers for some to-be-defined type and algorithm using Provider.getService and creates an instance via Provider.service.newInstance
* No option to have KDF and MAC come from different providers

Reply:
I'm not quite following the `Kdf.getInstance("HKDF", ...)` thought, though I also don't understand JCA. 😄

AIUI, `Kdf.getInstance` would be if we want to swap out the HKDF implementation itself (not sure why one would want that, but whatever), which we're not currently pursuing since JCA doesn't yet have a notion of KDFs as an algorithm. So in the meantime, we're just making a concrete HKDF class. That concrete class needs to call into the HMAC function at various points. My comment was just being a little puzzled at how involved calling the HMAC function was.

I.e., other than the `NoSuchAlgorithmException` issue, for which I don't have a great answer, it seems it could simply be...

```
    private final String hmacName;
    // Single argument ctor leaves provider null. providerName ctor fills uses Security.getProvider.
    private final Provider provider;

    private Mac getMac(byte[] key) throws InvalidKeyException, NoSuchAlgorithmException {
        Mac mac;
        if (provider != null) {
            mac = Mac.getInstance(hmacName, provider);
        } else {
            mac = Mac.getInstance(hmacName);
        }
        mac.init(new SecretKeySpec(key, "RAW"));
        return mac;
    }
```

Or we could even skip the provider-specific API because I doubt anyone actually cares. :-) But I don't understand JCA, so maybe there's a reason we don't want that... is doing a provider lookup on every `getMac` call expensive?

Anyway, consider this an optional thought, not a request to change things. I defer to you on JCA stuff since I don't understand it. (Why did Java make this so unusable and needlessly complicated...)

Reply:
> I'm not quite following the `Kdf.getInstance("HKDF", ...)` thought, though I also don't understand JCA. 😄

I was looking ahead, it seems like upstream are going to bundle a bunch of stuff under the Kdf type, including HKDF and their existing PBE algorithms.  The interface classes might turn out... interesting.

> My comment was just being a little puzzled at how involved calling the HMAC function was.

Yeah, when I wrote it I was thinking we should try very hard to always find a Conscrypt HMAC implementation unless the caller specified otherwise... But after discussion I agree, it's better just to trust people to set up the JCA Provider ordering.  In the default Android and OpenJDK use cases that'll "just work" unless people go out of their way to break it.  The workarounds I originally wrote don't really add value and they're just a bit too Rube Goldberg.

I _was_ tempted to try the ideas I suggested above, mainly as a trial run for splitting the HPKE stuff into a client API and backend SPI, but actually I'm just going to simplify it and skip the `Provider` stuff, at least for now.

> Or we could even skip the provider-specific API because I doubt anyone actually cares. :-) But I don't understand JCA, so maybe there's a reason we don't want that... is doing a provider lookup on every `getMac` call expensive?

It's more expensive than you'd like but in the global scheme of things it's not going to be a bottleneck.